src/audio/google/google_ctc_audio_processing.c

```diff
@@ -0,0 +1,369 @@
+	}
+
+	if (!config) {
+		comp_err(dev, "google_ctc_audio_processing_reconfigure(): Tuning config not set");
+		return -EINVAL;
+	}
+
+	comp_info(dev, "google_ctc_audio_processing_reconfigure(): New tuning config %p (%zu bytes)",
+		  config, size);
+
+	cd->reconfigure = false;
+
+	uint8_t *google_ctc_audio_processing_config;
+	size_t google_ctc_audio_processing_config_size;
+	int num_channels;
+	bool google_ctc_audio_processing_config_present;
+	bool num_channels_present;
+
+	GoogleCtcAudioProcessingParseSofConfigMessage(config, size,
+						      &google_ctc_audio_processing_config,
+						      &google_ctc_audio_processing_config_size,
+						      &num_channels,
+						      &google_ctc_audio_processing_config_present,
+						      &num_channels_present);
+
+	if (google_ctc_audio_processing_config_present) {
+		comp_info(dev,
+			  "google_ctc_audio_processing_reconfigure(): Applying config of size %zu bytes",
+			  google_ctc_audio_processing_config_size);
+
+		ret = GoogleCtcAudioProcessingReconfigure(cd->state,
+							  google_ctc_audio_processing_config,
+							  google_ctc_audio_processing_config_size);
+		if (ret) {
+			comp_err(dev, "GoogleCtcAudioProcessingReconfigure failed: %d",
+				 ret);
+			return ret;
+		}
+	}
+
+	if (num_channels_present) {
+		cd->num_channels = num_channels;
+		comp_info(dev,
+			  "google_ctc_audio_processing_reconfigure(): Applying num channels %d",
+			  cd->num_channels);
+	}
+
+	return 0;
+}
+
+static int google_ctc_audio_processing_prepare(struct processing_module *mod)
+{
+	struct comp_dev *dev = mod->dev;
+	struct google_ctc_audio_processing_comp_data *cd = module_get_private_data(mod);
+
+	comp_info(dev, "google_ctc_audio_processing_prepare()");
+
+	return 0;
+}
+
+static int google_ctc_audio_processing_reset(struct processing_module *mod)
+{
+	comp_dbg(mod->dev, "google_ctc_audio_processing_reset()");
+
+	return 0;
+}
+
+static int google_ctc_audio_processing_process(struct processing_module *mod,
+					       struct input_stream_buffer *input_buffers,
+					       int num_input_buffers,
+					       struct output_stream_buffer *output_buffers,
+					       int num_output_buffers)
+{
+	struct google_ctc_audio_processing_comp_data *cd = module_get_private_data(mod);
+	int16_t *src;
+	int16_t *dst;
+
+	struct input_stream_buffer *in_streamb;
+	struct output_stream_buffer *out_streamb;
+	struct audio_stream *in_stream, *out_stream;
+
+	comp_dbg(dev, "google_ctc_audio_processing_process()");
+
+	if (cd->reconfigure) {
+		ret = google_ctc_audio_processing_reconfigure(mod);
+		if (ret)
+			return ret;
+	}
+
+	in_streamb = &input_buffers[0];
+	in_stream = in_streamb->data;
+	src = audio_stream_get_rptr(in_stream);
+	out_streamb = &output_buffers[0];
+	out_stream = out_streamb->data;
+	dst = audio_stream_get_rptr(out_stream);
+
+	for (frame = 0; frame < cd->num_frames; frame++) {
+		for (channel = 0; channel < cd->num_channels; channel++) {
+			GoogleCtcAudioProcessingProcess(cd->state, src, dst);
```

Comment:
So... the interface to the CTC library is a 3-argument C function that needs to be called for Every Individual Sample in the stream?  Is that not a performance disaster?  Can't we rework it so it takes a buffer pointer?

Reply:
Changed to process for the whole chunk. Thanks for pointing this out.